File: dotnet/VedaDB/Wire/Vbp/VBPFrame.cs
// VedaDB .NET SDK — VBP wire layer
//
// 8-byte VBP frame header: 3B magic 'VDB' (0x56 0x44 0x42) + 4B LE payload_length + 1B seq.
// payload_length INCLUDES op + flags + body. Min payload = 2 (op + flags).
// Max payload = 64 MiB (matches Go reference).
//
// Wire layout (VBP_SPEC.md §2):
//   +--------+---------+-----+-----+-----+----------+...+
//   | 'VDB'  | len_le4 | seq | op  | flg |  body    |
//   +--------+---------+-----+-----+-----+----------+...+
//   | 3 B    | 4 B     | 1 B | 1 B | 1 B | (len-2)B |
//   +--------+---------+-----+-----+-----+----------+...+

using System;
using System.Buffers.Binary;

namespace VedaDB.Wire.Vbp
{
    public sealed class VBPFrame
    {
        public static readonly byte[] Magic = new byte[] { (byte)'V', (byte)'D', (byte)'B' };
        public const int MagicLen = 3;
        public const int LenLen = 4;
        public const int SeqLen = 1;
        public const int HdrLen = MagicLen + LenLen + SeqLen; // 8
        public const int OpLen = 1;
        public const int FlagsLen = 1;
        public const int OpFlagsLen = OpLen + FlagsLen; // 2
        public const int MaxFrameLen = 64 * 1024 * 1024; // 64 MiB

        public byte Seq { get; }
        public byte Op { get; }
        public byte Flags { get; }
        public byte[] Body { get; }

        public VBPFrame(byte seq, byte op, byte flags, byte[] body)
        {
            if (body == null) throw new ArgumentNullException(nameof(body));
            if (body.Length > MaxFrameLen - OpFlagsLen)
                throw new VBPOversizeException($"body too large: {body.Length}");
            Seq = seq;
            Op = op;
            Flags = flags;
            Body = body;
        }

        public int PayloadLength => OpFlagsLen + Body.Length;
        public int TotalLength => HdrLen + OpFlagsLen + Body.Length;

        /// <summary>Encode this frame to a fresh byte array (VBP_SPEC.md §2 wire layout).</summary>
        public byte[] Encode()
        {
            var outBuf = new byte[TotalLength];
            EncodeTo(outBuf, 0);
            return outBuf;
        }

        /// <summary>Encode into an existing buffer at <paramref name="offset"/>. Returns bytes written.</summary>
        public int EncodeTo(byte[] outBuf, int offset)
        {
            int p = offset;
            outBuf[p++] = Magic[0];
            outBuf[p++] = Magic[1];
            outBuf[p++] = Magic[2];
            int pl = PayloadLength;
            BinaryPrimitives.WriteInt32LittleEndian(outBuf.AsSpan(p, 4), pl);
            p += 4;
            outBuf[p++] = Seq;
            outBuf[p++] = Op;
            outBuf[p++] = Flags;
            if (Body.Length > 0)
            {
                Buffer.BlockCopy(Body, 0, outBuf, p, Body.Length);
                p += Body.Length;
            }
            return p - offset;
        }

        /// <summary>Decode a frame from <paramref name="buf"/> starting at <paramref name="offset"/>.</summary>
        public static VBPFrame Decode(byte[] buf, int offset)
        {
            if (buf == null) throw new ArgumentNullException(nameof(buf));
            if (buf.Length - offset < HdrLen)
                throw new VBPTruncatedException($"need at least {HdrLen} bytes");

            if (buf[offset] != Magic[0] || buf[offset + 1] != Magic[1] || buf[offset + 2] != Magic[2])
                throw new VBPBadMagicException($"bad magic at offset {offset}");

            int pl = BinaryPrimitives.ReadInt32LittleEndian(buf.AsSpan(offset + 3, 4));
            byte seq = buf[offset + 7];
            if (pl < OpFlagsLen)
                throw new VBPTruncatedException($"payload_length {pl} < {OpFlagsLen}");
            if (pl > MaxFrameLen)
                throw new VBPOversizeException($"payload_length {pl} > {MaxFrameLen}");

            int need = HdrLen + pl;
            if (buf.Length - offset < need)
                throw new VBPTruncatedException(
                    $"buffer truncated: need {need}, have {buf.Length - offset}");

            byte op = buf[offset + HdrLen];
            byte flags = buf[offset + HdrLen + 1];
            int bodyLen = pl - OpFlagsLen;
            byte[] body = new byte[bodyLen];
            if (bodyLen > 0)
                Buffer.BlockCopy(buf, offset + HdrLen + OpFlagsLen, body, 0, bodyLen);
            return new VBPFrame(seq, op, flags, body);
        }

        public override string ToString() =>
            $"Frame(seq={Seq}, op={VBPOpcodes.OpcodeName(Op)}, flags=0x{Flags:X2}, body_len={Body.Length})";

        public override bool Equals(object? obj)
        {
            if (obj is not VBPFrame other) return false;
            if (Seq != other.Seq || Op != other.Op || Flags != other.Flags) return false;
            if (Body.Length != other.Body.Length) return false;
            for (int i = 0; i < Body.Length; i++)
                if (Body[i] != other.Body[i]) return false;
            return true;
        }

        public override int GetHashCode() => Seq * 31 + Op;
    }
}

File: php/src/VedaDB/Wire/Vbp/VBPFrame.php
<?php
declare(strict_types=1);

namespace VedaDB\Wire\Vbp;

/**
 * VBP frame: 8-byte header (3B magic 'VDB' + 4B LE payload_length + 1B seq) + body.
 *
 * Wire layout (VBP_SPEC.md §2):
 *
 *     +--------+---------+-----+-----+-----+----------+...+
 *     | 'VDB'  | len_le4 | seq | op  | flg |  body    |
 *     +--------+---------+-----+-----+-----+----------+...+
 *     | 3 B    | 4 B     | 1 B | 1 B | 1 B | (len-2)B |
 *     +--------+---------+-----+-----+-----+----------+...+
 *
 * Magic is the literal ASCII bytes 0x56 0x44 0x42.
 * payload_length is op + flags + body size (little-endian u32).
 *
 * MAX_FRAME_LEN is 64 MiB (matches the Go reference).
 */
final class VBPFrame
{
    public const MAGIC = ['V', 'D', 'B']; // 0x56 0x44 0x42
    public const MAGIC_BYTES = "\x56\x44\x42";
    public const MAGIC_LEN = 3;
    public const LEN_LEN = 4;
    public const SEQ_LEN = 1;
    public const HDR_LEN = self::MAGIC_LEN + self::LEN_LEN + self::SEQ_LEN; // 8
    public const OP_LEN = 1;
    public const FLAGS_LEN = 1;
    public const OPFLAGS_LEN = self::OP_LEN + self::FLAGS_LEN; // 2
    public const MAX_FRAME_LEN = 64 * 1024 * 1024; // 64 MiB

    public function __construct(
        public readonly int $seq,
        public readonly int $op,
        public readonly int $flags,
        public readonly string $body,
    ) {
        self::checkRange('seq', $seq, 0, 0xFF);
        self::checkRange('op', $op, 0, 0xFF);
        self::checkRange('flags', $flags, 0, 0xFF);
        if (strlen($body) > self::MAX_FRAME_LEN - self::OPFLAGS_LEN) {
            throw new VBPProtocolException(
                VBPProtocolError::Oversize,
                'body too large: ' . strlen($body),
            );
        }
    }

    public function payloadLength(): int
    {
        return self::OPFLAGS_LEN + strlen($this->body);
    }

    public function totalLength(): int
    {
        return self::HDR_LEN + self::OPFLAGS_LEN + strlen($this->body);
    }

    /** Encode this frame to a fresh byte string. */
    public function encode(): string
    {
        $body = $this->body;
        $pl = $this->payloadLength();
        // Build: magic (3) + len_le4 (4) + seq (1) + op (1) + flags (1) + body
        return self::MAGIC_BYTES
            . pack('V', $pl)
            . chr($this->seq & 0xFF)
            . chr($this->op & 0xFF)
            . chr($this->flags & 0xFF)
            . $body;
    }

    /**
     * Decode a frame from $buf starting at $offset.
     *
     * @throws VBPProtocolException on bad magic, truncated, oversize, or out-of-range
     */
    public static function decode(string $buf, int $offset = 0): self
    {
        $len = strlen($buf) - $offset;
        if ($len < self::HDR_LEN) {
            throw new VBPProtocolException(
                VBPProtocolError::Truncated,
                'need at least ' . self::HDR_LEN . ' bytes, have ' . $len,
            );
        }
        if ($buf[$offset] !== 'V' || $buf[$offset + 1] !== 'D' || $buf[$offset + 2] !== 'B') {
            throw new VBPProtocolException(
                VBPProtocolError::BadMagic,
                'bad magic at offset ' . $offset,
            );
        }
        // Read 4-byte little-endian payload length.
        $unpacked = unpack('Vpl', substr($buf, $offset + self::MAGIC_LEN, self::LEN_LEN));
        if ($unpacked === false) {
            throw new VBPProtocolException(VBPProtocolError::Truncated, 'unpack failed');
        }
        $pl = $unpacked['pl'];
        $seq = ord($buf[$offset + self::MAGIC_LEN + self::LEN_LEN]) & 0xFF;
        if ($pl < self::OPFLAGS_LEN) {
            throw new VBPProtocolException(
                VBPProtocolError::Truncated,
                "payload_length $pl < " . self::OPFLAGS_LEN,
            );
        }
        if ($pl > self::MAX_FRAME_LEN) {
            throw new VBPProtocolException(
                VBPProtocolError::Oversize,
                "payload_length $pl > " . self::MAX_FRAME_LEN,
            );
        }
        $need = self::HDR_LEN + $pl;
        if (strlen($buf) - $offset < $need) {
            throw new VBPProtocolException(
                VBPProtocolError::Truncated,
                "buffer truncated: need $need, have " . (strlen($buf) - $offset),
            );
        }
        $op = ord($buf[$offset + self::HDR_LEN]) & 0xFF;
        $flags = ord($buf[$offset + self::HDR_LEN + 1]) & 0xFF;
        $bodyLen = $pl - self::OPFLAGS_LEN;
        $body = $bodyLen > 0
            ? substr($buf, $offset + self::HDR_LEN + self::OPFLAGS_LEN, $bodyLen)
            : '';
        return new self($seq, $op, $flags, $body);
    }

    private static function checkRange(string $name, int $v, int $lo, int $hi): void
    {
        if ($v < $lo || $v > $hi) {
            throw new \InvalidArgumentException("$name out of range: $v");
        }
    }

    public function __toString(): string
    {
        return sprintf(
            'Frame(seq=%d, op=0x%02x, flags=0x%02x, body_len=%d)',
            $this->seq,
            $this->op & 0xFF,
            $this->flags & 0xFF,
            strlen($this->body),
        );
    }
}
